File: app/src/main/java/com/example/enterprisehub/LoginActivity.java
package com.example.enterprisehub;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;

public class LoginActivity extends AppCompatActivity {

    private static final String PREFS_NAME = "EnterpriseHubPrefs";
    private static final String KEY_PIN = "user_pin";
    public static final String KEY_LAST_ACTIVE = "last_active";

    private EditText etPin;
    private TextView tvStatus;
    private boolean isSettingUp = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        etPin = findViewById(R.id.et_pin);
        tvStatus = findViewById(R.id.tv_status);
        Button btnConfirm = findViewById(R.id.btn_confirm);

        SharedPreferences prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        String savedPin = prefs.getString(KEY_PIN, null);

        if (savedPin == null) {
            isSettingUp = true;
            tvStatus.setText("Create New PIN (4 Digits)");
        } else {
            isSettingUp = false;
            tvStatus.setText("Enter PIN");
        }

        btnConfirm.setOnClickListener(v -> {
            String input = etPin.getText().toString();
            if (input.length() != 4) {
                Toast.makeText(this, "PIN must be 4 digits", Toast.LENGTH_SHORT).show();
                return;
            }

            if (isSettingUp) {
                prefs.edit().putString(KEY_PIN, input).apply();
                Toast.makeText(this, "PIN Set!", Toast.LENGTH_SHORT).show();
                launchMain();
            } else {
                if (input.equals(savedPin)) {
                    launchMain();
                } else {
                    etPin.setText("");
                    Toast.makeText(this, "Incorrect PIN", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void launchMain() {
        // Reset Inactivity Timer
        getSharedPreferences(PREFS_NAME, MODE_PRIVATE).edit().putLong(KEY_LAST_ACTIVE, System.currentTimeMillis()).apply();

        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/com/samsunghub/app/ui/LoginActivity.kt
package com.samsunghub.app.ui

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.GridLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.samsunghub.app.MainActivity
import com.samsunghub.app.R
import com.samsunghub.app.utils.UserPrefs

class LoginActivity : AppCompatActivity() {

    private var enteredPin = StringBuilder()
    private val dots = ArrayList<View>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (!UserPrefs.isPinSet(this)) {
            startMainActivity()
            return
        }

        setContentView(R.layout.activity_login)
        setupUI()

        findViewById<TextView>(R.id.tvForgot).setOnClickListener {
            android.app.AlertDialog.Builder(this)
                .setTitle("Forgot PIN?")
                .setMessage("To reset your PIN, you must clear the app data from Android Settings.\n\nSettings > Apps > SamsungHub > Storage > Clear Data")
                .setPositiveButton("OK", null)
                .show()
        }
    }

    private fun setupUI() {
        dots.add(findViewById(R.id.dot1))
        dots.add(findViewById(R.id.dot2))
        dots.add(findViewById(R.id.dot3))
        dots.add(findViewById(R.id.dot4))

        val grid = findViewById<GridLayout>(R.id.gridLayout)
        for (i in 0 until grid.childCount) {
            val child = grid.getChildAt(i)
            if (child is Button) {
                child.setOnClickListener { v ->
                    val tag = v.tag.toString()
                    when (tag) {
                        "CLR" -> clearPin()
                        "GO" -> validatePin()
                        else -> addDigit(tag)
                    }
                }
            }
        }
    }

    private fun addDigit(digit: String) {
        if (enteredPin.length < 4) {
            enteredPin.append(digit)
            updateDots()
            if (enteredPin.length == 4) {
                // Auto validate on 4th digit
                validatePin()
            }
        }
    }

    private fun clearPin() {
        enteredPin.setLength(0)
        updateDots()
    }

    private fun updateDots() {
        for (i in 0 until 4) {
            if (i < enteredPin.length) {
                dots[i].setBackgroundResource(R.drawable.pin_dot_filled_blue)
            } else {
                dots[i].setBackgroundResource(R.drawable.pin_dot_empty_blue)
            }
        }
    }

    private fun validatePin() {
        val savedPin = UserPrefs.getPin(this)
        if (enteredPin.toString() == savedPin) {
            startMainActivity()
        } else {
            // Shake Animation
            val dotsContainer = findViewById<View>(R.id.dotsContainer)
            dotsContainer.animate()
                .translationX(20f)
                .setDuration(50)
                .withEndAction {
                    dotsContainer.animate()
                        .translationX(-20f)
                        .setDuration(50)
                        .withEndAction {
                            dotsContainer.animate().translationX(0f).setDuration(50).start()
                        }
                        .start()
                }
                .start()

            Toast.makeText(this, "Incorrect PIN", Toast.LENGTH_SHORT).show()
            enteredPin.setLength(0)
            updateDots()
        }
    }

    private fun startMainActivity() {
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }
}
